File: tests/workers/test_facedetect.py
# Copyright 2021 Xilinx, Inc.
# Copyright 2022 Advanced Micro Devices, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import math

import numpy as np
import pytest

import amdinfer
import amdinfer.testing

from helper import root_path, run_benchmark


@pytest.mark.extensions(["aks", "vitis"])
@pytest.mark.fpgas("DPUCADF8H", 1)
@pytest.mark.usefixtures("load")
class TestInferImageFacedetectDPUCADF8H:
    """
    Test the facedetect worker
    """

    @staticmethod
    def get_config():
        model = ["CPlusPlus", "AksDetect"]
        parameters = amdinfer.ParameterMap(["model"], ["base64_decode"])

        aks_parameters = amdinfer.ParameterMap()
        aks_parameters.put("aks_graph_name", "facedetect")
        aks_parameters.put(
            "aks_graph",
            "${AKS_ROOT}/graph_zoo/graph_facedetect_u200_u250_amdinfer.json",
        )
        return (model, [parameters, aks_parameters])

    def send_request(self, request, check_asserts=True):
        """
        Sends the given request to the server and asserts common checks

        Args:
            request (InferenceRequest): Request to send to the server
            image (np.ndarray): Golden image to check against
            check_asserts (bool): Verify image against golden

        Returns:
            Response: Response as a dictionary
        """

        try:
            response = self.rest_client.modelInfer(self.endpoint, request)
        except amdinfer.ConnectionError:
            pytest.fail(
                "Connection to the amdinfer server ended without response!", False
            )

        num_inputs = len(request.getInputs())
        # for this picture and xmodel combination, we expect the following output
        gold_response_output = [
            -1,
            0.9937100410461426,
            266,
            78.728,
            158,
            170.800,
        ]

        if check_asserts:
            assert not response.isError(), response.getError()
            assert response.id == ""
            assert response.model == "facedetect"
            outputs = response.getOutputs()
            assert len(outputs) == num_inputs
            for index, output in enumerate(outputs):
                assert output.name == ""
                assert output.datatype == amdinfer.DataType.FP32
                assert output.parameters.empty()
                data = output.getFp32Data()
                num_boxes = int(len(data) / 6)
                assert output.shape == [num_boxes, 6]
                assert len(data) == len(
                    gold_response_output
                ), f"{len(data)} != {len(gold_response_output)}"
                for datum, golden in zip(data, gold_response_output):
                    # expect that the response values are within 5% of the golden
                    abs_error = abs(golden * 0.05)
                    assert math.isclose(datum, golden, abs_tol=abs_error)
        return response

    def construct_request(self, asTensor):
        image_path = amdinfer.testing.getPathToAsset("asset_girl-1867092_640.jpg")
        batch = 1
        images = [image_path] * batch
        request = amdinfer.ImageInferenceRequest(images, asTensor)

        return request

    def test_facedetect_dpucadf8h_1(self):
        """
        Send a request to facedetect as base64-encoded data
        """
        request = self.construct_request(False)
        self.send_request(request)

    @pytest.mark.benchmark(group="facedetect_dpucadf8h")
    def test_benchmark_facedetect_dpucadf8h_1(self, benchmark):
        request = self.construct_request(False)
        options = {
            "model": self.model,
            "parameters": self.parameters,
            "type": "rest (pytest)",
            "config": "N/A",
        }
        run_benchmark(
            benchmark, "facedetect_1", self.rest_client._infer, request, **options
        )

File: tests/workers/test_facedetect.cpp
// Copyright 2021 Xilinx, Inc.
// Copyright 2022 Advanced Micro Devices, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <cstdlib>  // for getenv, abs

#include "amdinfer/testing/get_path_to_asset.hpp"  // for getPathToAsset
#include "facedetect.hpp"                          // IWYU pragma: associated
#include "gtest/gtest.h"  // for Test, AssertionResult, EXPECT_EQ

const int gold_response_size = 6;
const float gold_response_output[gold_response_size] = {
  -1, 0.9937100410461426, 268, 78.728, 158, 170.800,
};

std::string prepareDirectory() {
  fs::path temp_dir =
    fs::temp_directory_path() / "amdinfer/tests/cpp/native/facedetect";
  fs::create_directories(temp_dir);
  const auto src_file =
    fs::path(amdinfer::getPathToAsset("asset_girl-1867092_640.jpg"));
  fs::copy_file(src_file, temp_dir / src_file.filename(),
                fs::copy_options::skip_existing);
  return temp_dir;
}

void dequeue_validate(FutureQueue& my_queue, int num_images) {
  std::future<amdinfer::InferenceResponse> element;
  for (int i = 0; i < num_images; i++) {
    my_queue.wait_dequeue(element);
    auto results = element.get();

    EXPECT_STREQ(results.getID().c_str(), "");
    EXPECT_STREQ(results.getModel().c_str(), "facedetect");
    auto outputs = results.getOutputs();
    EXPECT_EQ(outputs.size(), 1);
    for (auto& output : outputs) {
      const auto* data = static_cast<float*>(output.getData());
      auto size = output.getSize();
      EXPECT_STREQ(output.getName().c_str(), "");
      EXPECT_STREQ(output.getDatatype().str(), "FP32");
      auto* parameters = output.getParameters();
      ASSERT_NE(parameters, nullptr);
      EXPECT_TRUE(parameters->empty());
      auto num_boxes = gold_response_size / 6;
      auto shape = output.getShape();
      EXPECT_EQ(shape.size(), 2);
      EXPECT_EQ(shape[0], 6);
      EXPECT_EQ(shape[1], num_boxes);
      EXPECT_EQ(size, gold_response_size);
      for (size_t i = 0; i < gold_response_size; i++) {
        // expect that the response values are within 1% of the golden
        const float abs_error = std::abs(gold_response_output[i] * 0.05);
        EXPECT_NEAR(data[i], gold_response_output[i], abs_error);
      }
    }
  }
}

// @pytest.mark.extensions(["vitis"])
// @pytest.mark.fpgas("DPUCADF8H", 1)
TEST(Native, Facedetect) {
  amdinfer::Server server;
  amdinfer::NativeClient client;

  auto fpgas_exist = client.hasHardware("DPUCADF8H", 1);
  if (!fpgas_exist) {
    GTEST_SKIP() << "No FPGAs available";
  }

  auto path = prepareDirectory();
  auto worker_name = load(1);
  auto image_paths = getImages(path);
  auto num_images = image_paths.size();

  FutureQueue my_queue;
  run(image_paths, 1, worker_name, my_queue);

  dequeue_validate(my_queue, num_images);
}
